src/v/kafka/server/snc_quota_manager.h

```diff
@@ -58,6 +61,29 @@ class snc_quotas_probe {
     size_t _traffic_in = 0;
 };
 
+class snc_quota_context {
+public:
+    explicit snc_quota_context(std::optional<std::string_view> client_id)
+      : _client_id(client_id) {}
+
+private:
+    friend class snc_quota_manager;
```

Comment:
nit: it sort of looks like you could move snc_quota_context into snc_quota_manager.cc and then out here only deal with opaque `std::unique_ptr<ctx>` becaues it looks like all the call sites etc.. only ever pass in references to the manager itself.

Reply:
Indeed, I have missed that. Thanks!

Reply:
Having had a closer look, this makes things worse actually. Exposing unique_ptr to an incomplete class and giving away ownership of these objects requires custom deleter, which has to be complete at the point of unique_ptr declaration. If I place the deleter into `snc_quota_manager.h`, this header needs to be included into `connection_context.h` -> header bloat; putting it into a separate file is an overkill. With the current approach though, `snc_quota_manager.h` is not included into any header file, so I suppose the contention is pretty good.

Reply:
You can usually forward declare the class that the `unique_ptr` holds, and then declare a destructor for the class that holds the `unique_ptr`, and the in the implementation file, you can define the destructor as default.

Reply:
Yeah that's a good pattern for a regular pimpl, but this case is not a pimpl actually. It's like this:

`inner.h`:
```c++
class inner_context;

class inner {
  unique_ptr<inner_context> _context;
public:
  unique_ptr<inner_context> get_context();
  ~inner();
};
```
`inner.cc`:
```c++
class inner_context {};
inner::~inner() = default;
```
`outer.h`:
```c++
class outer {
  unique_ptr<inner> _inner;
  unique_ptr<inner_context> _inner_context;
public:
  void f () {
     _inner_context = _inner->get_context();
  }
  ~outer();
};
```

For that to work, we need either of:
* `inner_context` to be complete in both `inner.cc` and `outer.cc` (where dtors for the corresponding classes are defined)
* attach a custom deleter to the `unique_ptr<inner,inner_deleter>`, and the deleter must be complete at all declarations of `unique_ptr<inner,inner_deleter>`, i.e. in both `inner.h` and `outer.h`.

The first option is apparently cleaner, causes less header dependencies, and simpler. It only needs `inner_context` to be defined in `inner.h`, and there is not much to hide there actually.